flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQDeserializationSchema.java

```diff
@@ -0,0 +1,90 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.rabbitmq;
+
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.util.Collector;
+
+import com.rabbitmq.client.AMQP;
+import com.rabbitmq.client.Envelope;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+
+/**
+ * Interface for the set of methods required to parse an RMQ delivery.
+ * @param <T> The output type of the {@link RMQSource}
+ */
+public abstract class RMQDeserializationSchema<T> implements  Serializable, ResultTypeQueryable<T> {
+
+	/**
+	 * Initialization method for the schema. It is called before the actual working methods
+	 * {@link #deserialize} and thus suitable for one time setup work.
+	 *
+	 * <p>The provided {@link DeserializationSchema.InitializationContext} can be used to access additional features
+	 * such as e.g.
+	 * registering user metrics.
+	 *
+	 * @param context Contextual information that can be used during initialization.
+	 */
+	void open(DeserializationSchema.InitializationContext context) throws Exception {
+
+	}
+
+	/**
+	 * This method takes all the RabbitMQ delivery information supplied by the client extract the data and pass it to
+	 * the collector.
+	 *
+	 * <p><b>NOTICE:</b> The implementation of this method can call {@link RMQCollector#setMessageIdentifiers} with
+	 * a custom correlation ID and delivery tag if checkpointing and UseCorrelationID (in the RMQSource constructor)
+	 * were enabled
+	 * the {@link RMQSource}.
+	 * @param envelope an AMQP {@link Envelope}.
+	 * @param properties the {@link AMQP.BasicProperties} of the message.
+	 * @param body the message itself as a byte array.
+	 * @param collector the {@link RMQCollector} that will collect the data.
+	 * @throws IOException When the body of the message can't be parsed
+	 */
+	abstract void deserialize(Envelope envelope,
+								AMQP.BasicProperties properties,
+								byte[] body,
+								RMQCollector<T> collector) throws IOException;
+
+	/**
+	 * Method to decide whether the element signals the end of the stream. If
+	 * true is returned the element won't be emitted.
+	 *
+	 * @param nextElement The element to test for the end-of-stream signal.
+	 * @return True, if the element signals end of stream, false otherwise.
+	 */
+	abstract boolean isEndOfStream(T nextElement);
+
+	/**
+	 * Special collector for RMQ messages.
+	 *
+	 * <p>It extends the {@link Collector} to give the ability to collect more than 1 message and the ability to set
+	 * the message correlationId and deliveryTag.
+	 */
+	public interface RMQCollector<T> extends Collector<T> {
+		void collect(List<T> records);
+
+		void setMessageIdentifiers(String correlationId, long deliveryTag);
```

Comment:
Could you add a javadoc to this method? I think it is quite important to do so. Moreover even I personally would appreciate explanation what is the difference between `correlationId` and `deliveryTag` I am not too familiar with RMQ and personally don't know what is the difference.

Reply:
I also thought we could add `boolean` as the result type, to determine if we saw this message or not. That way we could potentially skip deserialization of the record in the schema, as we do in current master. (We do not even call `deserialize` if the message is a duplicate, but we probably can not achieve that with the Collector)